### ISSUE TITLE
chmod doesn't work?

### ISSUE THREAD

**Issue by gencoimports** [Microsoft/WSL#81]
chmod doesn't seem to edit permissions. The command registers fine in bash, but the changes are not made to the actual files.

**Comment by dethoma** [comment]
Is this inside /mnt/ NTFS drives? We explicitly do not change Windows ACLs on them today based on chown/chmod. These commands work with full fidelity on the Linux filesystem in the lxss folder.

**Comment by gencoimports** [comment]
I'm a bit of a noob, so bear with me. But if I understand this correctly, chmod won't work on any files contained in the /mnt/[ntfs dive] partitions? BUT will work fine in the /.../lxss folders and/or other linux systems?

**Comment by dethoma** [comment]
Yes you are correct, it works inside the lxss folders where we have /home folder etc and we adhere to strict Linux rules. For files in the /mnt/c/* locations we do not want to modify Windows ACLs.

**Comment by vyv03354** [comment]
I understand you do not want to map permissions to NTFS ACLs. If WSL does it at all, it should map POSIX ACLs rather than traditional permissions.

But could you at least map writable permissions to read-only attributes under /mnt/c/? It will fail to write read-only files anyway, but Linux apps can't tell it will fail until it actually tries to write the file because the permission pretends to be writable.

**Comment by justinSelf** [comment]
Is there any plan in the future to allow chmod to "work" on /mnt/c?

There are several tools that create conventions around the permissions of a file to determine how to handle it.
 Ansible, for one.

**Comment by sunilmut** [comment]
We will be tracking\prioritizing which features to light up next by looking at the feedback provided at the wpdev portal. We appreciate your feedback and please make sure to provide it through the wpdev portal as well.

https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo

**Comment by AnneTheAgile** [comment]
Cross-reference; user voice request for ansible as master. 
https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13741026-ansible

**Comment by bakeiro** [comment]
i think this is bad structured, example: for install "x" app i must have 755 permision, okay so try to use chmod but only works on lxss directory so... move the app from windows to lxss directory to use chmod buuut from linux i simply can´t see files when comes from windows directory only when are created from linux.So whats next? i simply cant use chmod in /mnt/c and simply cant see files in lxss directory

**Comment by bakeiro** [comment]
i think this is bad structured, example: for install "x" app i must have 755 permision, okay so try to use chmod but only works on lxss directory so... move the app from windows to lxss directory to use chmod buuut from linux i simply can´t see files when comes from windows directory only when are created from linux.So whats next? i simply cant use chmod in /mnt/c and simply cant see files in lxss directory

**Comment by bitcrazed** [comment]
This isn't a priority for us and we have no plans to support this ask in the appreciable future. Closing for now.

If you'd like us to reconsider, please find/upvote this issue on [UserVoice](https://aka.ms/winbashuv) and share a little background on your scenarios.

**Comment by mufumbo** [comment]
many things that people use every-day won't ever work:
```

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0777 for '/mnt/c/Users/test/web/web.pem' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.
bad permissions: ignore key: /mnt/c/Users/test/web/web.pem
```

Any advice, rather than duplicating our files in a filesystem that is difficult to edit?

**Comment by erdemece** [comment]
If this is ubuntu on bash truly for helping developers to do more on windows this is really must do.

I use vagrant on windows and I can work in www folder on my windows folder system. So, I don't have to use terminal to write my code. I use sublime text and have Projects folder and I keep all my www virtual server folders in there. this makes it easier to develop. 

You can't do this with ubuntu on bash. this alone makes it no developer friendly at all.

**Comment by aseering** [comment]
@erdemece -- I do that all the time; it works great for me.  What specific problems are you seeing?  Please include (by copy&paste) the exact command-line command that you are running and the exact text of the error message that you are seeing.

For what it's worth, I as a WSL user have heard a lot of people complaining about this.  In my personal experience, around 15-20% of them have a very real use case, something that they can't do on WSL right now.  Those are real people with real use cases, so in my opinion this should absolutely get fixed.  But for 80-85% of the use cases I've heard, devs actually can do exactly what they want to do with either a very minor change to their workflow / a simple trick that they just didn't know about, or by changing a setting in some application, or sometimes WSL is actually working just fine and they just forgot to configure some piece of their dev environment on the new machine.

**Comment by computergeek125** [comment]
Example use case: SSH on both WSL and something else like Cygwin or PuTTY.  I can symlink my Windows .ssh directory into the WSL system (`cd ~; ln -s /mnt/c/Users/me/.ssh`.  I don't have my Windows box in front of me so I can't double-check that path).  With this setup, my SSH config and keys would be identical and I don't have to copy duplicate keys to each host.

Solution idea: perhaps instead of just globally setting NTFS permissions to 755, you could instead set the NTFS permissions to the current umask value.  With a umask of 755, it would be identical to the current behavior.  But, you could then set a umask of 700 to allow things like ssh to work.

**Comment by mufumbo** [comment]
@aseering I'm not sure how proficient you are with bash or linux systems. Please, see my use case above. 

One of the most trivial things you might want to do with bash is to use a program called "ssh" to open other linux boxes on the cloud. With the current windows-on-ubuntu setup, you can't use ssh properly, since you can't use chmod command to configure your ssh server keys. 

Hope it helps you to understand. thanks!

**Comment by aseering** [comment]
@mufumbo -- yep, your use case definitely makes sense to me.  Ditto for @computergeek125 ; it sounds like a very similar use case.  I use ssh all the time; I'm quite familiar :-)

As a workaround, could you simply have two copies of your SSH keys, one in a Windows-accessible directory and one in Linux?  Or do you change them often enough, or have enough of them, that this would be problematic?  If (for example) you have one key and you rotate it once per year, will the amortized cost of copying it to two places be prohibitive?  Of course, if you have lots of keys and/or use a key-management tool, then you're right that there is (to my knowledge) no great workaround right now.

I didn't mean to refer to you in my original comment.  I meant to refer specifically to @erdemece , whose concern seemed to be (if I understand it correctly) developing a Web app using Windows tools while serving it via Apache.  Apache is, generally speaking, happy to serve any files that it has access to and that are within its DocumentRoot (or that Apache is otherwise configured to serve).  If Apache isn't seeing the files, I would ask whether Apache is pointing directly at DrvFs or if it's pointing at a symlink.  In the latter case, I'd ask whether Apache has FollowSymlinks set correctly.  If it's pointing directly at DrvFs, I would ask whether Apache's own (Allow/Deny) permissions are set up to allow it to serve from that directory.  I can keep going for a while with the decision tree here -- basically, @erdemece 's particular report was very broad; the limited information provided so far could be pointing to any of several WSL limitations as well as a variety of ordinary system misconfigurations.  I just want a more-precise bug report :-)

**Comment by aseering** [comment]
@computergeek125 -- that's an interesting idea, but, a concern:  Linux permissions have two aspects.  The umask corresponds to the second aspect -- what is the file's owner/group allowed to do with the file?  The first is, who is the file's owner?  DrvFs doesn't track file ownership either; everything is owned by root.  If everything in DrvFs inherited root's default umask, then you would not be able to edit files in DrvFs as an ordinary Linux user, which I think would be frustrating to a lot of people.  And while most (not all!) WSL installations do have a default non-root user account, it's common to create additional users.  So, if WSL picked one user account to own the files, this would create problems for the other user accounts.

I suppose DrvFs could report that files are always owned by the "current user" -- whatever process is asking, it checks what user that process is running as, and reports files as being owned by that user.  But that seems to me like a risky level of magic...

**Comment by computergeek125** [comment]
Risky magic indeed.  I temporarily forgot that aspect.

**Comment by mufumbo** [comment]
@aseering yes, i tried those work-arounds, but in the end of the day the idea was to use windows for UI management and ubuntu as bash. I found it painful/impossible to have to deal with filesystems that can't communicate. (it was easier to install linux in a virtualbox machine and use bash from there, which is too painful too)

Do you think it would be possible for the ubuntu subsystem to have it's custom chmod's for windows FS in memory? That would solve many problems.

**Comment by aseering** [comment]
@mufumbo -- I'm not a WSL developer; I'm just a user.  (WSL devs have the "Member" or "Owner" tags on their posts.)  So, as to whether the permissions could be kept in memory (presumably in lxss somewhere so that they would be scoped to the current user) -- I'm sure it would be technically feasible, as long as you don't have too many files, but I'm not in a position to do it :-)

As a user, I'm trying to do something that's very similar to my reading of your description above, and it works great for me.  Small up-front cost; zero ongoing cost.  Vastly easier for me than a VM.  But it clearly doesn't work well for you.  So it seems to me that there's some difference that explains why things work for me and not for you.  Given that there fundamentally isn't one single simple clear way to map Linux permissions to DrvFs (do they affect Windows permissions?, do different Windows users' bash installs see the permissions?, etc), I think there's a reasonable chance that whatever you're hitting could be relevant to the design decisions that the WSL team might decide to make in the future.

**Comment by mufumbo** [comment]
@aseering thanks so much for your help and emphaty! 

I was so excited when I saw BashOnWindows that I bought a 32gb ram $3300 machine and decided to try windows. Unfortunately, it's not there yet and I have to go back to the lame apple hardware.

**Comment by hendorog** [comment]
Simple use case: 
Without chmod support it isn't possible to login to an Amazon instance. That requires ssh -i support.
I use the same script on OSX and linux, but it doesn't work on Windows because of this limitation.

**Comment by aseering** [comment]
@hendorog --  SSH keys are actually discussed a bunch in this thread already, including workarounds.  I use AWS all the time from WSL by placing my keys in `~/.ssh`.  Could you elaborate on what about your particular situation makes those workarounds difficult for you?  For example, do you need to store your keys alongside your script, and it would be annoying to (for example) place them in `~/.ssh` and symlink them to the directory of your script?  (ssh checks file permissions, not symlink permissions.)

( @mufumbo -- no problem; sorry I can't actually fix this.  Ooh, that's a nice machine.  It's probably poor form on a Windows forum, but I'm tempted to suggest making a [Hackintosh](http://www.hackintosh.com/) :-) )

**Comment by hendorog** [comment]
@aseering Yep the workarounds work, but workarounds are just that. The issue is that scripts using AWS would all need to be modified to support Windows, or symlinks used etc. 

ssh is just such a fundamental tool in unix world, AWS is very widely used, and key based login is more the norm than the exception now. 

Also, off the top of my head I thought from many years ago NTFS was posix compatible. So I thought implementing chmod/chown would be possible. Implementation left as an exercise for the reader of course :)

**Comment by matthewrk** [comment]
Another use case is web developers running webapps under WSL but editing in Windows, which I assume is extremely common unless I'm doing something wrong/weird. To do this the source has to be under /mnt/c which is all root:root. So I either have to run the webapp as root or my app has no write access to its working directory.

**Comment by aseering** [comment]
@hendorog -- one challenge with using NTFS POSIX ACLs is, who would own the file?  NTFS ACLs are global to all users on the machine.  But each Windows user account has its own separate set of Linux users.

(I for one would love to see real POSIX ACLs be used here, and I can think of a few different ways to do that.  Just stating one place in which the concepts don't map in the way that one might like.)

@matthewrk -- it's true that the files are owned by root, but it's not necessarily true that only the owner of a file can edit that file.  In this specific case, the files have permissions `0777`, which means that all users can edit the file regardless of who owns them.

**Comment by matthewrk** [comment]
@aseering you're correct! It was slightly blurred for me because I'm using rsync to workaround issue #873 which wasn't copying the permissions. With that knowledge though I should be able to fix it, thanks very much.

**Comment by hendorog** [comment]
@aseering -- there would need to be a mapping between a linux user and a windows user, which I think is how the posix interface was supposed to work?

**Comment by JuicyPasta** [comment]
I've found a solution! 

Basically the problem is NTFS or exFAT does not have any notion of POSIX inodes, or any way to store a file permissions. The solution is to create an EXT4 partition (I did it with a bootable ubuntu flash drive) and mount it on windows with something like ext2fsd (very easy to set up). After you do that everything magically works which is very surprising but ill take it :)

Notice I am able to set permissions and the owner isn't root

![permissions](https://cloud.githubusercontent.com/assets/7462530/25068182/159cd512-2210-11e7-8e3c-87e4387353ef.png)

**Comment by davidmaxwaterman** [comment]
As a side note, this issue caused me a problem just now. I have been using WSL for some time now, and have been working entirely within WSL, so the files had correct ownership/permissions. However, I wanted to try out VSCode, but to use that I needed to move all my files into /mnt/c. I did that using `rsync -a` specifically to maintain the file attributes...the files copied fine, of course, but all ended up as root.root and 0777, which was unexpected.
This wouldn't be a problem, except it was a git clone and now `git status` tells me that every file has been changed, and I've done other relevant changes...unfortunately, it doesn't say *what* has changed and I have to do a `git diff` to see that it is just the file mode that has changed. I have ended up using `git diff | grep -v '^(old\|new) mode'` to help me see the real changes.
In summary, it is quite irritating - best to do a commit right after the files are moved, and then do more development changes.

**Comment by hippyjim** [comment]
@aseering none of the "workarounds" here fix the problem, so please dson't asume they do. I copied my ssh keys from their original windows location into wherever my ~/.ssh directory really is, but chmod still did nothing.

**Comment by AviBueno** [comment]
Having the same issue after a 'git stash' operation on files under /mnt/c/...
The changes are stashed but the permissions change (from 644 to 777 in my case) and the files seem like they are not stashed.

- 'git diff' shows 
    ```
    old mode 100644
    new mode 100755
    ```
- 'git reset --hard' doesn't have any affect on the files.
- 'chmod 644 ...' doesn't have an affect either.
- It is now impossible to do 'git stash pop' or even 'git stash apply' because of:
    > error: Your local changes to the following files would be overwritten by merge

I installed WSL a few days ago hoping to replace my cygwin setup, but for various reasons, chmod being one of them, the feeling is that WSL is not yet polished enough.

The question is: Why was this issue closed? it's is not resolved.

**Comment by levidehaan** [comment]
Yeah I am also curious why this issue is closed.

I am experiencing the similar issues, it'd be nice to have a resolution, especially since this is being touted as the new hawtness, and a reason i considered trying windows again.

or let us know that this will never be fixed (it has been a year already since this was reported) so we can move on and stop sinking time into WSL.

**Comment by aseering** [comment]
So, the underlying issue here is that DrvFs is simply exposing the Windows NTFS filesystem at /mnt/c/; NTFS is not typically used with Linux permissions, and WSL doesn't yet expose them at all.  I don't know why this is currently closed; I'll let the WSL team comment on that.

I will add that Windows git doesn't necessarily handle file permissions very well either.  The reason that you don't see it is that, as I understand it, the Windows client will do the equivalent of `git config core.fileMode false` automatically.  Unfortunately, the official Linux git client is not clever enough to do this automatically if needed, but you can run that command yourself if needed.  This is also an issue if you use Windows filesystems from real Linux, so while WSL ought to improve this in my opinion, I also think it would be fair to ask git to fix their Linux client to work as well as their Windows client in this scenario.

(And, yes, WSL is still explicitly in beta.  As a result, the intended expectation is that there will be rough edges like this.  This particular issue has been around ~forever, and I for one would like to see it fixed, but permissions are challenging to really do right, and WSL users have so far voted it as a [relatively low-importance issue](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/filters/top).  If you do care about it, I suggest that you either get engaged and upvote things that you care about, or, frankly, put WSL aside for another release or two and come back when it has built more stuff and is ready to give you the experience that you expect.  It is still evolving very quickly.)

**Comment by mufumbo** [comment]
There's no sense for this issue to be closed, please re-open it. 
Who we need to contact at Microsoft to make the change and get craving users what they want?

**Comment by benhillis** [comment]
Agreed with the above comments, this issue should not have been closed. File system support and performance is an area that we are currently spending a lot of effort to improve. We have some great improvements in the pipeline but they are in progress so it's hard to say for sure when they will land. 

Reopening this issue.

**Comment by davidmaxwaterman** [comment]
`git config core.fileMode false` helped me, fwiw.

**Comment by Entroper** [comment]
This one bit me today.  I've been using WSL to test NodeJS apps.  The code lives on my Windows filesystem, and I edit it with VS Code, and have a soft link to the code directory from lxss.  Everything's been working great, except today I couldn't add write permissions to a deployment script.  Git (running from WSL) didn't check in the file with execute permissions, and there was no way to add the permission from Bash.  And obviously `filemode false` wouldn't help here.

**Comment by Absurdo** [comment]
@davidmaxwaterman. Why rsync instead a symlink ?

**Comment by davidmaxwaterman** [comment]
@Absurdo I presume you're referring to my attempt to move my files from inside WSL to the /mnt/c filesystem? I didn't use symlink under the assumption that Windows wouldn't know what a linux symlink is and wouldn't be able to 'follow' it into the WSL filesystem. If a symlink from Windows into WSL would work, then surely I could just run VSCode directly on the WSL files without moving them at all.
Did I misunderstand your question?

**Comment by meiskam** [comment]
having confusing troubles with permissions brought me to this issue, would love a fix

**Comment by marcthe12** [comment]
Well could we look into ntfs-3g. I does the reverse, ntfs on linux. They have a User map.

**Comment by tavor** [comment]
Hope to save somebody a Google search. 

Came here to figure out why chmod 400 with bash/ubuntu on Windows seemed to not work. Did another Google search and found this:
https://mikethecanuck.wordpress.com/2016/09/27/why-doesnt-chmod-under-bash-on-ubuntu-on-windows-10-actually-take/

My setup is exactly how he described it. Stuff under my /mnt drive was owned by root. In home, I had things own by me.

My key used to be in a /mnt directory, and I wasn't able to get chmod 400 to take. I copied the key to ~/, chmod worked. scp'ing to my ec2 instance now works. That chmod didn't "work" was discouraging. Still, doesn't seem like I'm the only that missed this, and, if this works for more folks, I'd say it'd make sense to somehow make this consequence more obvious or make things easier for people setting stuff up, if it's truly lazy/bad setup that causes this.

I can believe that I was lazy during my initial setup of ubuntu/bash on Windows & that would hurt me, but I don't remember what I did now why things are working. My Linux knowledge is usually enough to get by, but it's not much more than that.

**Comment by bitcrazed** [comment]
Just posted a reply to the article above:

> Nice to see you're making progress, but it is important to understand the challenges in mapping the different file permission semantics between Linux DACs and NT ACLs:

> DAC is a much simpler permissions model than ACLs and DAC doesn't support many of the permission semantics that ACLs provide for.

> When you access files stored in your Windows filesystem via the /mnt/... folder, remember that WSL is expressing your **effective-calculated** permissions based on the credentials of the user & currently running process (I.e. Bash running as the logged-in user). For example, if you `ll /mnt/` you'll notice that 'Program Files' is not writeable, because the current user is not given write permissions to this folder by default.

> If you want to express specific Linux DAC permissions on a folder and/or files, then you'll need to copy those files into the Linux side of your filesystem where we DO maintain full DAC permission support. However, doing this prevents Windows tools from being able to access those files ... at least right now (Dec 2017). 

> We ARE working on improving this scenario, by exposing the Linux filesystem to Windows & Windows tools, but for now you'll need to be aware of the above and choose the most appropriate path. 

Let me know if you have any further questions on this matter.

**Comment by greggman** [comment]
I think I'm mis-understanding how 

     git config core.fileMode false

Is a solution to the git issues. 

     git config core.fileMode false

only tells git to ignore the executable flag. The other RW permission still get messed up and make it impossible manage git projects under /mnt/c so that I can edit with windows editors

What am I missing? Is there some other way to make git ignore permission changes?

**Comment by sunilmut** [comment]
With the DrvFS metadata support in [17063](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17063), `chmod` should now work (and stick) on DrvFS.

**Comment by tara-raj** [comment]
Marking this closed as per https://github.com/Microsoft/WSL/issues/81#issuecomment-356081660

**Comment by zanona** [comment]
Thanks, guys. I tested, and it worked well under `sudo`.
Although I need to open Ubuntu, stop all processes that are using `/mnt/c` and then restart with 
```bash
sudo umount /mnt/c
sudo mount -t drvfs C: /mnt/c -o metadata
```
This needs to be done every time an Ubuntu terminal is opened.

By any change will this be a default in the future, or any way to make this persistent across sessions?

Thanks again.

**Comment by onomatopellan** [comment]
@zanona If you are on build 17134 you can use wsl.conf. https://blogs.msdn.microsoft.com/commandline/2018/02/07/automatically-configuring-wsl/

**Comment by monkeysuffrage** [comment]
Wait, what goes in wsl.conf? To do @zanona's thing automatically?

**Comment by onomatopellan** [comment]
@monkeysuffrage with a linux editor create a /etc/wsl.conf file with this content
```
[automount]
enabled = true
options = "metadata"
mountFsTab = false
```
Close the Ubuntu window and open again. This will enable metadata for every fixed drive.

**Comment by dragGH102** [comment]
I have tried (build 1803) both - with and without - the wsl.conf suggested by @onomatopellan but chmod just doesn't work.

Note I'm using Debian, not Ubuntu but they're fundamentally the same afaik...

**Comment by onomatopellan** [comment]
@dragGH102 Just tried on Debian WSL and it works. What's your output of `cat /etc/wsl.conf` ?

**Comment by dragGH102** [comment]
@onomatopellan weird indeed. There was a typo but even after changing it to

```
[automount]
enabled = true
options = "metadata"
mountFsTab = false
```

(output of cat)

I still get the issue. I have restarted Debian app (not Windows - unsure whether that's necessary)

**Comment by onomatopellan** [comment]
@dragGH102 When wsl.conf is applied you can see the metadata option on `mount` output. You need to close every WSL open window and on Task Manager kill every "init" picoprocess you find. Seems some people need to reboot Windows in order to get wsl.conf working but I think it's because some WSL picoprocess remains on memory.

**Comment by dragGH102** [comment]
@onomatopellan thanks for clarifying. I can confirm that after a restart, it works

**Comment by Hi-Angel** [comment]
> If you'd like us to reconsider, please find/upvote this issue on UserVoice and share a little background on your scenarios.

@bitcrazed what was the point of asking to create a poll if you just closed the poll without providing a solution, nor you re-opened this issue?

**Comment by Hi-Angel** [comment]
Ah, nvm, per last comments in this issue it seems to be solved. Probably worth commenting on the user-voice page as well, sorry for the noise.

**Comment by theAkito** [comment]
# To Everyone Arriving here from Google

To clarify and summarize the solution to the original problem of this issue, I will explain how to apply the fix **without restarting your computer**.

1. Open `/etc/wsl.conf` with your favourite text editor.
2. Paste the following into it:
```
[automount]
enabled = true
options = "metadata"
mountFsTab = false
```
3. Close **all** WSL windows!
4. Open PowerShell.
5. List your WSL distributions:
```
wsl --list
```
6. Pick the one you want to fix. We select "Debian" as an example.
7. Terminate the Debian WSL instance:
```
wsl --terminate Debian
```
8. Open a new Debian WSL window. It should load for a couple of seconds, which already indicates that the whole WSL was terminated, before.
9. The fix should be applied now. If it still does not work, repeat all above steps and make absolutely sure, that no WSL windows remain open when applying the fix.
